pkg/controller/certificaterequests/ca/ca.go

```diff
@@ -17,24 +17,119 @@ limitations under the License.
 	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
 	"github.com/jetstack/cert-manager/pkg/controller/certificaterequests"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util"
+	"github.com/jetstack/cert-manager/pkg/util/kube"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
 )
 
 const (
 	CRControllerName = "certificaterequests-issuer-ca"
 )
 
+type CA struct {
+	// used to record Events about resources to the API
+	recorder record.EventRecorder
+
+	issuerOptions controllerpkg.IssuerOptions
+	secretsLister corelisters.SecretLister
+	helper        issuer.Helper
+}
+
 func init() {
 	// create certificate request controller for ca issuer
 	controllerpkg.Register(CRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
-		controller := certificaterequests.New(apiutil.IssuerCA)
+		ca := NewCA(ctx)
+
+		controller := certificaterequests.New(apiutil.IssuerCA, ca)
 
 		c, err := controllerpkg.New(ctx, CRControllerName, controller)
 		if err != nil {
 			return nil, err
 		}
+
 		return c.Run, nil
 	})
 }
+
+func NewCA(ctx *controllerpkg.Context) *CA {
+	return &CA{
+		recorder:      ctx.Recorder,
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		helper: issuer.NewHelper(
+			ctx.SharedInformerFactory.Certmanager().V1alpha1().Issuers().Lister(),
+			ctx.SharedInformerFactory.Certmanager().V1alpha1().ClusterIssuers().Lister(),
+		),
+	}
+}
+
+func (c *CA) Sign(ctx context.Context, cr *v1alpha1.CertificateRequest) (*issuer.IssueResponse, error) {
+	log := logf.FromContext(ctx, "sign")
+
+	issuer, err := c.helper.GetGenericIssuer(cr.Spec.IssuerRef, cr.Namespace)
+	if k8sErrors.IsNotFound(err) {
+		apiutil.SetCertificateRequestCondition(cr, v1alpha1.CertificateRequestConditionReady,
+			v1alpha1.ConditionFalse, v1alpha1.CertificateRequestReasonPending,
+			fmt.Sprintf("Referenced %s not found", util.IssuerKind(cr.Spec.IssuerRef)))
+
+		c.recorder.Event(cr, corev1.EventTypeWarning, v1alpha1.CertificateRequestReasonPending, err.Error())
+
+		log.WithValues(
+			logf.RelatedResourceNameKey, cr.Spec.IssuerRef.Name,
+			logf.RelatedResourceKindKey, cr.Spec.IssuerRef.Kind,
+		).Error(err, "failed to find referenced issuer")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	resourceNamespace := c.issuerOptions.ResourceNamespace(issuer)
+
+	// get a copy of the CA certificate named on the Issuer
+	caCerts, caKey, err := kube.SecretTLSKeyPair(ctx, c.secretsLister, resourceNamespace, issuer.GetSpec().CA.SecretName)
+	if k8sErrors.IsNotFound(err) {
+		log := logf.WithRelatedResourceName(log, issuer.GetSpec().CA.SecretName, resourceNamespace, "Secret")
+		log.Info("error getting signing CA for Issuer")
+
+		c.recorder.Event(cr, corev1.EventTypeWarning, v1alpha1.CertificateRequestReasonPending, err.Error())
+	}
+	if err != nil {
+		return nil, err
```

Comment:
Again, we shouldn't be returning an error here if the referenced `secretName` doesn't exist. Instead we should watch on changes to the Issuer (as the Ready condition will be updated once the Secret exists/is valid)